File: src/main/java/myblog/blog/category/domain/Category.java
package myblog.blog.category.domain;


import lombok.Builder;
import lombok.Getter;

import myblog.blog.shared.BasicEntity;
import myblog.blog.article.domain.Article;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

/*
    - 카테고리 엔티티
*/
@Entity
@Getter
@SequenceGenerator(
        name = "CATEGORY_SEQ_GENERATOR",
        sequenceName = "CATEGORY_SEQ",
        initialValue = 0, allocationSize = 50)
public class Category extends BasicEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "CATEGORY_SEQ_GENERATOR")
    @Column(name = "category_id")
    private Long id;

    @Column(nullable = false, unique = true, length = 20)
    private String title;

    @OneToMany(mappedBy = "category")
    private List<Article> articles = new ArrayList<>();

    @Column(nullable = false)
    private int tier;

    private int pSortNum;
    private int cSortNum;

    // 셀프조인
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parents_id")
    private Category parents;

    @OneToMany(mappedBy = "parents")
    private List<Category> child = new ArrayList<>();

    @Builder
    public Category(String title, Category parents, int tier, int pSortNum, int cSortNum) {
        this.title = title;
        this.parents = parents;
        this.tier = tier;
        this.pSortNum = pSortNum;
        this.cSortNum = cSortNum;
    }

    protected Category() {
    }

    @Override
    public String toString() {
        return title;
    }

    // 도메인 비지니스 로직

    /*
        - 카테고리 더티체킹 업데이트
    */
    public void updateCategory(String title, int tier, int pSortNum, int cSortNum, Category parents){
        this.title = title;
        this.tier = tier;
        this.pSortNum = pSortNum;
        this.cSortNum = cSortNum;
        this.parents = parents;
    }

}

File: service/blog/blog-core/src/main/kotlin/kr/co/jiniaslog/blog/domain/category/Category.kt
package kr.co.jiniaslog.blog.domain.category

import jakarta.persistence.AttributeOverride
import jakarta.persistence.Column
import jakarta.persistence.EmbeddedId
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import kr.co.jiniaslog.shared.adapter.out.rdb.JpaAggregate

/**
 * 카테고리 어그리게이트 루트
 *
 * - 구현 편의상 부모 카테고리와 자식카테고리 2계층으로만 구현한다
 */
@Entity
class Category(
    id: CategoryId,
    categoryTitle: CategoryTitle,
    sortingPoint: Int,
) : JpaAggregate<CategoryId>() {
    @EmbeddedId
    @AttributeOverride(
        column = Column(name = "id"),
        name = "value",
    )
    override val entityId: CategoryId = id

    @AttributeOverride(
        column = Column(name = "title"),
        name = "value",
    )
    var categoryTitle: CategoryTitle = categoryTitle
        private set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    var parent: Category? = null
        private set

    @OneToMany(mappedBy = "parent", fetch = FetchType.LAZY)
    var children: MutableList<Category> = mutableListOf()
        private set

    @Column(name = "sorting_point")
    var sortingPoint: Int = sortingPoint
        private set

    val isChild: Boolean
        get() = parent != null

    init {
        require(sortingPoint >= 0) { "sortingPoint는 0 이상이어야 합니다" }
    }

    fun changeParent(parent: Category?) {
        this.parent = parent
        parent?.children?.add(this)
    }

    fun sync(
        categoryTitle: CategoryTitle,
        sortingPoint: Int,
        parent: Category?,
    ) {
        require(sortingPoint >= 0) { "sortingPoint는 0 이상이어야 합니다" }
        this.categoryTitle = categoryTitle
        this.sortingPoint = sortingPoint
        this.parent = parent
    }
}
